python/private/python.bzl

```
@@ -78,6 +78,55 @@ def parse_modules(*, module_ctx, _fail = fail):
 
     config = _get_toolchain_config(modules = module_ctx.modules, _fail = _fail)
 
+    default_python_version = None
+    for mod in module_ctx.modules:
+        defaults_attr_structs = _create_defaults_attr_structs(mod = mod)
+        default_python_version_env = None
+        default_python_version_file = None
+
+        # Only the root module and rules_python are allowed to specify the default
+        # toolchain for a couple reasons:
+        # * It prevents submodules from specifying different defaults and only
+        #   one of them winning.
+        # * rules_python needs to set a soft default in case the root module doesn't,
+        #   e.g. if the root module doesn't use Python itself.
+        # * The root module is allowed to override the rules_python default.
+        if mod.is_root or (mod.name == "rules_python" and not default_python_version):
+            for defaults_attr in defaults_attr_structs:
+                default_python_version = _one_or_the_same(
+                    default_python_version,
+                    defaults_attr.python_version,
+                    onerror = _fail_multiple_defaults_python_version,
+                )
+                default_python_version_env = _one_or_the_same(
+                    default_python_version_env,
+                    defaults_attr.python_version_env,
+                    onerror = _fail_multiple_defaults_python_version_env,
+                )
+                default_python_version_file = _one_or_the_same(
+                    default_python_version_file,
+                    defaults_attr.python_version_file,
+                    onerror = _fail_multiple_defaults_python_version_file,
+                )
+            if default_python_version_file:
+                default_python_version = _one_or_the_same(
+                    default_python_version,
+                    module_ctx.read(default_python_version_file).strip(),
+                )
+            if default_python_version_env:
+                # Bazel version 7.1.0 and later support module_ctx.getenv(name, default):
+                # When building incrementally, any change to the value of the variable
+                # named by `name` will cause this repository to be re-fetched.
+                if "getenv" in dir(module_ctx):
+                    getenv = module_ctx.getenv
+                else:
+                    getenv = module_ctx.os.environ.get
+                default_python_version = getenv(default_python_version_env, default_python_version)
+    if not default_python_version:
+        fallback_python_version_file = module_ctx.path("@@//:.python-version")
+        if fallback_python_version_file.exists:
+            default_python_version = module_ctx.read(fallback_python_version_file).strip()
```

```suggestion
        if fallback_python_version_file.exists:
            default_python_version = module_ctx.read(fallback_python_version_file).strip()
```

We should ensure that the watch is done for the version file if we decide to read. I remember that `module_ctx.path` no longer watches things by default, but I can't remember if `module_ctx.read` does or not. Maybe  would be safer to pass `watch="yes"` to the read?

https://bazel.build/versions/8.1.0/rules/lib/builtins/module_ctx#read